### ISSUE TITLE
Deleting stashes causes SEGVs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3438]
Migrated from [rt.perl.org#5851](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5851) (status was 'resolved')

Searchable as RT5851$

**Comment by p5pRT** [comment]
### From @AlanBurlison



If you delete a stash and then try to call something it it you get a SEGV\.
Hardly surprising I suppose\, but perhaps something should be done about it
sometime\.

$ perl \-e 'sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
Segmentation Fault\(coredump\)

$ pstack core
core 'core' of 103834&#8203;:  /usr/local/bin/perl \-e sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
 ff27ca34 Perl\_hv\_fetch \(2e9a0\, ffbee6f0\, b\, 0\, a2411bd1\, 2d948\) \+ 16c
 ff229874 Perl\_gv\_fetchpv \(b\, 0\, 3a\, 0\, 2d348\, ff30ac24\) \+ 160
 ff229430 Perl\_gv\_stashpvn \(ff2f37e8\, ffbee850\, 0\, ff2f37e8\, 9\, 3a535550\) \+ 60
 ff228d6c Perl\_gv\_fetchmeth \(a\, 36174\, 7\, 0\, ff30ac24\, 3606c\) \+ 4ac
 ff229054 Perl\_gv\_fetchmethod\_autoload \(3606c\, 3606c\, 7\, ff30ac24\, ff2f86f8\, 1\) \+ 160
 ff295660 Perl\_sv\_clear \(2faf4\, 1000\, 34770\, ff0b6000\, 22994\, ff0419f8\) \+ b4
 ff296010 Perl\_sv\_free \(360e4\, ff30ac24\, 0\, 1c\, 2296c\, ff276190\) \+ 1c0
 ff22b220 Perl\_gp\_free \(36a08\, 360f0\, d\, 600d\, 22994\, ff0419f8\) \+ 130
 ff295bc8 Perl\_sv\_clear \(0\, 1000\, c\, 0\, 2296c\, ff27ef34\) \+ 61c
 ff296010 Perl\_sv\_free \(360f0\, ff30ac24\, 360f0\, ff30ac24\, 2e968\, 32f2c\) \+ 1c0
 ff27e5b0 Perl\_hv\_free\_ent \(2d948\, 34080\, ff3a0414\, 0\, ff3e2698\, ff20e503\) \+ 94
 ff27e7ec S\_hfreeentries \(1f\, 363c8\, 7c\, 1f\, 33db0\, 2d948\) \+ 4c
 ff27e840 Perl\_hv\_undef \(2e9a0\, 2d948\, 2000000b\, 0\, 0\, 0\) \+ 14
 ff29efb8 Perl\_pp\_undef \(2d49c\, 2d4a0\, ff30ac24\, 2faf4\, 2d948\, 2faf4\) \+ 13c
 ff2822d8 Perl\_runops\_standard \(ff30ac24\, 2d49c\, ffbeeecc\, 0\, ff3e2698\, 11572\) \+ 44
 ff222940 S\_run\_body \(1\, ff31b494\, ff30ac24\, ff317090\, ff30ac24\, ff31b648\) \+ 1dc
 ff222564 perl\_run \(2d4d0\, 11b94\, 3\, ffbeefec\, 0\, 3\) \+ 2a0
 00011b44 main     \(0\, 2d330\, 2cd08\, ffbeefec\, 0\, 0\) \+ 9c
 00011a80 \_start   \(0\, 0\, 0\, 0\, 0\, 0\) \+ b8

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by alanbur at Tue May 23 19:20:13 BST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.8, archname=sun4-solaris-64int
    uname='sunos mower 5.8 generic sun4u sparc sunw,ultra-60 '
    config_args='-dsOE -Dprefix=/usr/local -Dinstallprefix=/usr/local -Dsiteprefix=/usr/local -Doptimize=-xO3 -Duseshrplib -Uusemymalloc -Ubincompat5005 -Duse64bitint -Accflags=-DDL_UNLOAD_ALL_AT_EXIT'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=define use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-xO3', gccversion=
    cppflags='-DDL_UNLOAD_ALL_AT_EXIT'
    ccflags ='-DDL_UNLOAD_ALL_AT_EXIT -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/opt/SUNWspro/SC5.0/lib '
    libpth=/opt/SUNWspro/SC5.0/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='  -R /usr/local/lib/perl5/5.6.0/sun4-solaris-64int/CORE'
    cccdlflags='-KPIC', lddlflags='-G -L/opt/SUNWspro/SC5.0/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/sun4-solaris-64int
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/sun4-solaris-64int
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home1/homedir/alanbur
    LANG=C
    LANGUAGE (unset)
    LC_COLLATE=en_GB.ISO8859-1
    LC_CTYPE=en_GB.ISO8859-1
    LC_MESSAGES=C
    LC_MONETARY=en_GB.ISO8859-1
    LC_NUMERIC=en_GB.ISO8859-1
    LC_TIME=en_GB.ISO8859-1
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home1/homedir/alanbur/bin:/home1/homedir/alanbur/bin/sparc:/usr/bin:/usr/sbin:/usr/local/bin:/usr/dt/bin:/usr/openwin/bin:/usr/ccs/bin:/opt/SUNWspro-local/bin:/usr/dist/exe:/usr/dist/local/exe:/home1/homedir/alanbur/perforce:/home1/homedir/oracle/product/8.1.6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @tux

On Mon\, 19 Feb 2001 11&#8203;:23&#8203;:53 GMT\, alanbur@&#8203;mower\.uk\.sun\.com wrote&#8203;:
> 
> This is a bug report for perl from alanbur@&#8203;mower\.uk\.sun\.com\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> If you delete a stash and then try to call something it it you get a SEGV\.
> Hardly surprising I suppose\, but perhaps something should be done about it
> sometime\.
> 
> $ perl \-e 'sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
> Segmentation Fault\(coredump\)

Should we allow to delete/undef %&#8203;:: \(%main&#8203;::\) anytime at all?

**Comment by p5pRT** [comment]
### From @simoncozens

On Mon\, Feb 19\, 2001 at 05&#8203;:12&#8203;:27PM \+0100\, H\. Merijn Brand wrote&#8203;:
> > $ perl \-e 'sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
> > Segmentation Fault\(coredump\)
> 
> Should we allow to delete/undef %&#8203;:: \(%main&#8203;::\) anytime at all?

I would say yes\, on the "stupid things possible" principle or whatever
it is\. But I don't know what decent behaviour in the above case would be\.
I think it would be easy enough to croak with a "sub has gone away" type
rror\.

Simon

**Comment by p5pRT** [comment]
### From @AlanBurlison

"H\.Merijn Brand" wrote&#8203;:

> Should we allow to delete/undef %&#8203;:: \(%main&#8203;::\) anytime at all?

I suspect undef of any stash with a sub in it that is subsequently
called will do the same thing \- main&#8203;:: isn't special in that respect\.  I
guess what \*should\* happen is that stash deletion should invalidate the
method cache\, and a subsequent attempt to call the sub should fail with
'sub not defined' or somesuch\.
 
Alan Burlison

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Alan Burlison \<Alan\.Burlison@&#8203;uk\.sun\.com> writes&#8203;:
> "H\.Merijn Brand" wrote&#8203;:
>>
>> Should we allow to delete/undef %&#8203;:: \(%main&#8203;::\) anytime at all?
>
> I suspect undef of any stash with a sub in it that is subsequently
> called will do the same thing \- main&#8203;:: isn't special in that respect\.  I
> guess what \*should\* happen is that stash deletion should invalidate the
> method cache\, and a subsequent attempt to call the sub should fail with
> 'sub not defined' or somesuch\.

Well I get a stack dump with ACCVIO \(equiv of SEGV\) from just&#8203;:

&nbsp;    $ perl \-e"sub foo\{print 'hi';\} undef %main&#8203;::;"

\(note no attempt to call the sub\)

But
&nbsp;    $ perl \-e"sub BAR&#8203;::foo\{print 'hi';\} undef %BAR&#8203;::; BAR&#8203;::foo\(\);"
&nbsp;    hi
seems not only to not ACCVIO\, but also fails to delete the sub \(also
true if the definition of BAR&#8203;::foo is in a module with "package BAR;\.\.\."
type definitions\.\)

**Comment by p5pRT** [comment]
### From @AlanBurlison

Charles Lane wrote&#8203;:

> But
>     $ perl \-e"sub BAR&#8203;::foo\{print 'hi';\} undef %BAR&#8203;::; BAR&#8203;::foo\(\);"
>     hi
> seems not only to not ACCVIO\, but also fails to delete the sub \(also
> true if the definition of BAR&#8203;::foo is in a module with "package BAR;\.\.\."
> type definitions\.\)

This is because of the refcount loop between a CV \(code block\) and its
parent GV \(glob\)\.  At present subs are effectively immortal because of
this loop\.  If you look back over my other recent mails you'll see I've
wittered on at length about this\.

I'm currently working on fixing this\, but the fix has uncovered a really
nasty bug in the scoping of %^H\, which I'm currently trying to track
down\.

Alan Burlison

**Comment by p5pRT** [comment]
### From gt4556a@prism.gatech.edu



undef\(\)ing the symbol table of package main causes coredumps\.  The problem has
been reproduced on at least two version of Perl \(5\.00505 and 5\.6\.1\, both on
SunOS\) and two people \(on different machines\)\.

Test program&#8203;:

sub aaa \{\};
undef %&#8203;::;

Note that the coredump doesn't occur for certain names of the above sub\.
Examples&#8203;:

\* a
\* aa
\* b
\* \[try letters in between?\]
\* z
\* za \- zh

There are certainly more names\.  That don't trigger the coredump\.  Defining a
variable instead of a sub before undef\(\)ing %&#8203;:: gives no coredump either\.

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Configured by willday at Fri Apr  2 11:27:26 EST 1999.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos rom 5.6 generic_105181-12 sun4u sparc sunw,ultra-1 '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='-fast', gccversion=
    cppflags=''
    ccflags =''
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-fast -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -lgdbm -ldb -ldl -lm -lc -lcrypt
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00503:
    /fuse4/47/gt4556a/perl/sun4-solaris
    /fuse4/47/gt4556a/perl
    /usr/local/lib/perl5/5.00503/sun4-solaris
    /usr/local/lib/perl5/5.00503
    /usr/local/lib/perl5/site_perl/5.005/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/fuse4/47/gt4556a
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/fuse4/47/gt4556a/local/bin:/usr/bin:/usr/ccs/bin:/opt/SUNWspro/bin:/usr/local/bin::/tnt1/38/gtwreck/bin:/fuse4/47/gt4556a/bin:/fuse4/47/gt4556a/local/bin:/fuse4/47/gt4556a/nmh/bin:/fuse4/47/gt4556a/nmh/lib:/tnt1/38/gtwreck/bin:/fuse4/47/gt4556a/bin:/fuse4/47/gt4556a/local/bin:/fuse4/47/gt4556a/nmh/bin:/fuse4/47/gt4556a/nmh/lib
    PERL5LIB=/fuse4/47/gt4556a/perl
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From jeff@zeroclue.com


This is a bug report for perl from jeff@&#8203;zeroclue\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.4\.




---
Perl crashes \(seg fault\) when emptying the main namespace in any of the special blocks\.  I\.E\.

jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'CHECK\{%&#8203;::=\(\)\}'
Segmentation fault
jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'BEGIN\{%&#8203;::=\(\)\}'
Segmentation fault
jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'END\{%&#8203;::=\(\)\}'
Segmentation fault
jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'INIT\{%&#8203;::=\(\)\}'
Segmentation fault





---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.4&#8203;:

Configured by Debian Project at Sun Sep 26 12&#8203;:11&#8203;:30 CEST 2004\.

Summary of my perl5 \(revision 5 version 8 subversion 4\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.8\-1\-686\, archname=i386\-linux\-thread\-multi
&nbsp;    uname='linux cachaca 2\.6\.8\-1\-686 \#1 sat aug 28 14&#8203;:11&#8203;:39 edt 2004 i686 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=i386\-linux \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.8 \-Darchlib=/usr/lib/perl/5\.8 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.8\.4 \-Dsitearch=/usr/local/lib/perl/5\.8\.4 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Uusesfio \-Uusenm \-Duseshrplib \-Dlibperl=libperl\.so\.5\.8\.4 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.3\.4 \(Debian 1&#8203;:3\.3\.4\-12\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.3\.2\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.8\.4
&nbsp;    gnulibc\_version='2\.3\.2'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.4&#8203;:
&nbsp;    /home/jeff/dts/base/lib
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.8\.4
&nbsp;    /usr/local/share/perl/5\.8\.4
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.8
&nbsp;    /usr/share/perl/5\.8
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.4&#8203;:
&nbsp;    HOME=/home/jeff
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=en\_US\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/jeff/bin&#8203;:/usr/local/java/bin&#8203;:/home/jeff/dts/base/bin&#8203;:/usr/local/sbin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/home/jeff/bin&#8203;:/usr/local/java/bin&#8203;:/home/jeff/dts/base/bin&#8203;:/usr/local/sbin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games
&nbsp;    PERL5LIB=/home/jeff/dts/base/lib
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From mjtg@cam.ac.uk

Jeff Lavallee \<jeff@&#8203;zeroclue\.com> wrote
> Perl crashes \(seg fault\) when emptying the main namespace in any of
> the special blocks\.  I\.E\.
>
> jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'CHECK\{%&#8203;::=\(\)\}'
> Segmentation fault

This is a variation on an old chestnut\.   To help track it down\, I note

perl \-we 'BEGIN \{ delete $&#8203;::\{q\(@&#8203;\)\}\}'
Segmentation Fault

whereas

perl \-w
BEGIN \{
delete @&#8203;&#8203;::\{
"ENV"\,
"CORE\\&#8203;:\\&#8203;:"\,
"\\\+"\,
"STDERR"\,
"STDOUT"\,
"\\\<none\\>\\&#8203;:\\&#8203;:"\,
"\\""\,
"\_\\\<perlmain\\\.c"\,
"DynaLoader\\&#8203;:\\&#8203;:"\,
"main\\&#8203;:\\&#8203;:"\,
"Internals\\&#8203;:\\&#8203;:"\,
"DATA"\,
"ARGV"\,
"\\/"\,
"Regexp\\&#8203;:\\&#8203;:"\,
"\\x12"\,
"UNIVERSAL\\&#8203;:\\&#8203;:"\,
"0"\,
"attributes\\&#8203;:\\&#8203;:"\,
"STDIN"\,
"\_\\\<perlio\\\.c"\,
"IO\\&#8203;:\\&#8203;:"\,
"INC"\, 
"\\x18"\, 
"DB\\&#8203;:\\&#8203;:"\, 
"\_\\\<universal\\\.c"\, 
"PerlIO\\&#8203;:\\&#8203;:"\, 
"\_"\, 
"\_\\\<\\\-"\, 
"\\\-"\, 
"stdout"\, 
"\\x08"\, 
"utf8\\&#8203;:\\&#8203;:"\, 
"\\$"\, 
"\_\\\<xsutils\\\.c"\, 
"stderr"\, 
"stdin"\, 
\}\};
\_\_END\_\_

doesn't fail\.    That's all the other keys found by doing

perl \-wle 'print foreach keys %&#8203;::'

So $@&#8203; is the thing to stare at\.

And as an extra data point \(in case you haven't already guessed\)

perl \-we 'eval q\(delete $&#8203;::\{q\(@&#8203;\)\}\)'
Segmentation Fault


Mike Guy

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Jeff Lavallee \(via RT\) wrote&#8203;:
> Perl crashes \(seg fault\) when emptying the main namespace in any of the special blocks\.  I\.E\.
> 
> jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'CHECK\{%&#8203;::=\(\)\}'
> Segmentation fault
> jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'BEGIN\{%&#8203;::=\(\)\}'
> Segmentation fault
> jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'END\{%&#8203;::=\(\)\}'
> Segmentation fault
> jeff@&#8203;pokey&#8203;:\~ $ perl \-e 'INIT\{%&#8203;::=\(\)\}'
> Segmentation fault

And likewise

&nbsp;  $ perl \-e 'eval\{%&#8203;::=\(\)\}'
&nbsp;  Segmentation fault

There is a whole class of problems that come from this\.
In this case\, it's $@&#8203; \(or\, internally\, PL\_errgv\) being accessed \_after\_
it has been deleted\.

A more general\, but unclean\, fix\, would be to avoid deleting any GV in
%&#8203;:: that has the GvMULTI flag set\. More clean\, add a flag for "global
immortal" variables that should never be deleted from %&#8203;::\. I don't know
where to find room for flags\, though\. Opinions ?

\-\- 
\-\-Metempsychosis?
\-\-Yes\. Who's he when he's at home?
&nbsp;    \-\- Ulysses

**Comment by p5pRT** [comment]
### From @JohnPeacock

Rafael Garcia\-Suarez wrote&#8203;:

> A more general\, but unclean\, fix\, would be to avoid deleting any GV in
> %&#8203;:: that has the GvMULTI flag set\. More clean\, add a flag for "global
> immortal" variables that should never be deleted from %&#8203;::\. I don't know
> where to find room for flags\, though\. Opinions ?

I got bitten by this a few times too \(while spelunking in the core to 
add an overloaded object\, re&#8203;: version objects\)\.  If the problem case is 
$@&#8203;/PL\_errgv\, can't we just refuse to delete that out of $&#8203;:: instead of 
trying to make a more general solution?  The only time that can safely 
be nuked \(AFAICT\) is during final cleanup when Perl itself is shutting down\.

John

\-\- 
John Peacock
Director of Information Research and Technology
Rowman & Littlefield Publishing Group
4501 Forbes Boulevard
Suite H
Lanham\, MD  20706
301\-459\-3366 x\.5010
fax 301\-429\-5748

**Comment by p5pRT** [comment]
### From @JohnPeacock

Rafael Garcia\-Suarez wrote&#8203;:

> I remember also fixing a similar problem with STDERR\, or was it STDOUT\.
> I suspect that other problems may be lurking\.
> 

Yes\, it is likely that the same problem exists with both of those as 
well \(although they may already be fixed\)\.  I would go so far as to 
speculate that the problem cases are all ones where Perl itself is 
communicating with the outside world\.

I'm just suggesting that there is probably a minimum number of 
"specials" which need to be protected by name\, rather than trying to 
forge a more generic solution\.  Even if you did find a flag bit to play 
with \(and I don't know where one might be found\, cause I looked for the 
v\-string mess\)\, you'd still have to identify which gv's needed protecting\.

John

\-\- 
John Peacock
Director of Information Research and Technology
Rowman & Littlefield Publishing Group
4501 Forbes Boulevard
Suite H
Lanham\, MD  20706
301\-459\-3366 x\.5010
fax 301\-429\-5748

**Comment by p5pRT** [comment]
### From @rgs

John Peacock wrote&#8203;:
> 
> I got bitten by this a few times too \(while spelunking in the core to 
> add an overloaded object\, re&#8203;: version objects\)\.  If the problem case is 
> $@&#8203;/PL\_errgv\, can't we just refuse to delete that out of $&#8203;:: instead of 
> trying to make a more general solution?  The only time that can safely 
> be nuked \(AFAICT\) is during final cleanup when Perl itself is shutting down\.

I remember also fixing a similar problem with STDERR\, or was it STDOUT\.
I suspect that other problems may be lurking\.

\-\- 
Q&#8203;: How many extreme programmers does it take to change a lightbulb?
A&#8203;: Two \(Firstly to write the regression tests\, then to declare that it's a
&nbsp;   hardware problem\)  \-\- Nicholas Clark in p5p

**Comment by p5pRT** [comment]
### From @rgs

John Peacock wrote&#8203;:
> Yes\, it is likely that the same problem exists with both of those as 
> well \(although they may already be fixed\)\.  I would go so far as to 
> speculate that the problem cases are all ones where Perl itself is 
> communicating with the outside world\.
> 
> I'm just suggesting that there is probably a minimum number of 
> "specials" which need to be protected by name\, rather than trying to 
> forge a more generic solution\.  Even if you did find a flag bit to play 
> with \(and I don't know where one might be found\, cause I looked for the 
> v\-string mess\)\, you'd still have to identify which gv's needed protecting\.

That's why I was suggesting to use GvMULTI for this purpose : globals
vars that are not used only once\, i\.e\. real built\-in globals and
imported variables\. Although deleting imported variables from %&#8203;:: might
be justifiable\. Maybe :\)

\-\- 
Grepping the source is good for the soul\. \-\- the perldebguts manpage

**Comment by p5pRT** [comment]
### From @JohnPeacock

Rafael Garcia\-Suarez wrote&#8203;:

> That's why I was suggesting to use GvMULTI for this purpose : globals
> vars that are not used only once\, i\.e\. real built\-in globals and
> imported variables\. Although deleting imported variables from %&#8203;:: might
> be justifiable\. Maybe :\)
> 

I don't know\, I think it's pretty daft to nuke $&#8203;:: like that in the 
first place\.  It might be justifiable to just forbid the mass deletion 
itself and force anyone who wants to\, to walk the hash and delete the 
keys manually\.  Is that harder to code?

Anyone want to give an actual use\-case for clearing the $&#8203;:: in the first 
place???

John

\-\- 
John Peacock
Director of Information Research and Technology
Rowman & Littlefield Publishing Group
4501 Forbes Boulevard
Suite H
Lanham\, MD  20706
301\-459\-3366 x\.5010
fax 301\-429\-5748

**Comment by p5pRT** [comment]
### From @rgs

John Peacock wrote&#8203;:
> I don't know\, I think it's pretty daft to nuke $&#8203;:: like that in the 
> first place\.  It might be justifiable to just forbid the mass deletion 
> itself and force anyone who wants to\, to walk the hash and delete the 
> keys manually\.  Is that harder to code?

Won't help :

$ perl \-e 'eval\{delete$&#8203;::\{q\{@&#8203;\}\}\}'
Segmentation fault

> Anyone want to give an actual use\-case for clearing the $&#8203;:: in the first 
> place???

Clean up the namespace when you've imported lots of symbols ? If you happen
to re\-load things afterwards ?

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Dec 14\, 2004 at 12&#8203;:12&#8203;:12PM \-0500\, John Peacock wrote&#8203;:
> I don't know\, I think it's pretty daft to nuke $&#8203;:: like that in the 
> first place\.  It might be justifiable to just forbid the mass deletion 
> itself and force anyone who wants to\, to walk the hash and delete the 
> keys manually\.  Is that harder to code?

I submitted a patch a year or two back to forbid packagemainicide\, but
it was rejected\, presumably on the grounds that it was too much of a
special case\. There again\, %&#8203;:: deletion bugs seem to crop up with tedious
regularity\. IIRC correctly\, my patch stopped the obvious cases of
deletion\, but wouldn't stop pathological cases \(ie if you \*really\* tried
you could still do it\)\.

Dave\.

\-\- 
"Strange women lying in ponds distributing swords is no basis for a system
of government\. Supreme executive power derives from a mandate from the
masses\, not from some farcical aquatic ceremony\."
&nbsp;    \-\- Dennis \- Monty Python and the Holy Grail\.

**Comment by p5pRT** [comment]
### From @hvds

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;mandrakesoft\.com> wrote&#8203;:
:And likewise
:
:  $ perl \-e 'eval\{%&#8203;::=\(\)\}'
:  Segmentation fault
:
:There is a whole class of problems that come from this\.
:In this case\, it's $@&#8203; \(or\, internally\, PL\_errgv\) being accessed \_after\_
:it has been deleted\.
:
:A more general\, but unclean\, fix\, would be to avoid deleting any GV in
:%&#8203;:: that has the GvMULTI flag set\. More clean\, add a flag for "global
:immortal" variables that should never be deleted from %&#8203;::\. I don't know
:where to find room for flags\, though\. Opinions ?

I wonder whether an alternative route to fixing is to allow the deletion\,
but cope with it by ensuring that variables that need to exist are
revivified when required\.

I've no idea though how easy that would be to do\, nor whether blindly
assuming they'll exist saves enough time to warrant putting checks
elsewhere to ensure the assumption remains valid\.

Hugo

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Dec 15\, 2004 at 02&#8203;:51&#8203;:55PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> I wonder whether an alternative route to fixing is to allow the deletion\,
> but cope with it by ensuring that variables that need to exist are
> revivified when required\.
> 
> I've no idea though how easy that would be to do\, nor whether blindly
> assuming they'll exist saves enough time to warrant putting checks
> elsewhere to ensure the assumption remains valid\.

Or by moving out the code that creates them into a separate function\,
still called from the regular setup\, but arranging for it also to be called
as a special case just after %&#8203;:: is deleted\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @davidnicol

 perl \-e ' CHECK\{local \*&#8203;::; %&#8203;::=\(\)\}' 

doesn't crash\.  Maybe a better solution is to localize \*&#8203;:: and let it fall
out of scope\, instead of trying to clear it?  Although writing around this
would undeniably be a headache\.
&nbsp;  


\-\- 
David L Nicol
You're striving for harmony\, and\, if you try to take
 too much\, you'll come to grief\.  \-\- Michael Redmond

**Comment by p5pRT** [comment]
### From @smpeters

> \[gt4556a@&#8203;prism\.gatech\.edu \- Tue Jun 19 09&#8203;:30&#8203;:12 2001\]&#8203;:
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> undef\(\)ing the symbol table of package main causes coredumps\.  The
>    problem has
> been reproduced on at least two version of Perl \(5\.00505 and 5\.6\.1\,
>    both on
> SunOS\) and two people \(on different machines\)\.
> 
> Test program&#8203;:
> 
> sub aaa \{\};
> undef %&#8203;::;
> 
> Note that the coredump doesn't occur for certain names of the above
>    sub\.
> Examples&#8203;:
> 
> \* a
> \* aa
> \* b
> \* \[try letters in between?\]
> \* z
> \* za \- zh
> 
> There are certainly more names\.  That don't trigger the coredump\.
>    Defining a
> variable instead of a sub before undef\(\)ing %&#8203;:: gives no coredump
>    either\.
> 

As of 5\.6\.2\, this does still core dump; however\, as of at least 5\.8\.6\,
this does not core dump\, so this ticket is resolved\.

**Comment by p5pRT** [comment]
### From @smpeters

> \[alanbur@&#8203;mower\.uk\.sun\.com \- Sun Feb 18 19&#8203;:24&#8203;:39 2001\]&#8203;:
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> If you delete a stash and then try to call something it it you get a
>    SEGV\.
> Hardly surprising I suppose\, but perhaps something should be done
>    about it
> sometime\.
> 
> $ perl \-e 'sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
> Segmentation Fault\(coredump\)
> 
> $ pstack core
> core 'core' of 103834&#8203;:  /usr/local/bin/perl \-e sub bang \{ print
>    "bang\!\\n"; \} undef %main&#8203;::; bang;'
>  ff27ca34 Perl\_hv\_fetch \(2e9a0\, ffbee6f0\, b\, 0\, a2411bd1\, 2d948\) \+ 16c
>  ff229874 Perl\_gv\_fetchpv \(b\, 0\, 3a\, 0\, 2d348\, ff30ac24\) \+ 160
>  ff229430 Perl\_gv\_stashpvn \(ff2f37e8\, ffbee850\, 0\, ff2f37e8\, 9\,
>    3a535550\) \+ 60
>  ff228d6c Perl\_gv\_fetchmeth \(a\, 36174\, 7\, 0\, ff30ac24\, 3606c\) \+ 4ac
>  ff229054 Perl\_gv\_fetchmethod\_autoload \(3606c\, 3606c\, 7\, ff30ac24\,
>    ff2f86f8\, 1\) \+ 160
>  ff295660 Perl\_sv\_clear \(2faf4\, 1000\, 34770\, ff0b6000\, 22994\,
>    ff0419f8\) \+ b4
>  ff296010 Perl\_sv\_free \(360e4\, ff30ac24\, 0\, 1c\, 2296c\, ff276190\) \+ 1c0
>  ff22b220 Perl\_gp\_free \(36a08\, 360f0\, d\, 600d\, 22994\, ff0419f8\) \+ 130
>  ff295bc8 Perl\_sv\_clear \(0\, 1000\, c\, 0\, 2296c\, ff27ef34\) \+ 61c
>  ff296010 Perl\_sv\_free \(360f0\, ff30ac24\, 360f0\, ff30ac24\, 2e968\,
>    32f2c\) \+ 1c0
>  ff27e5b0 Perl\_hv\_free\_ent \(2d948\, 34080\, ff3a0414\, 0\, ff3e2698\,
>    ff20e503\) \+ 94
>  ff27e7ec S\_hfreeentries \(1f\, 363c8\, 7c\, 1f\, 33db0\, 2d948\) \+ 4c
>  ff27e840 Perl\_hv\_undef \(2e9a0\, 2d948\, 2000000b\, 0\, 0\, 0\) \+ 14
>  ff29efb8 Perl\_pp\_undef \(2d49c\, 2d4a0\, ff30ac24\, 2faf4\, 2d948\, 2faf4\)
>    \+ 13c
>  ff2822d8 Perl\_runops\_standard \(ff30ac24\, 2d49c\, ffbeeecc\, 0\,
>    ff3e2698\, 11572\) \+ 44
>  ff222940 S\_run\_body \(1\, ff31b494\, ff30ac24\, ff317090\, ff30ac24\,
>    ff31b648\) \+ 1dc
>  ff222564 perl\_run \(2d4d0\, 11b94\, 3\, ffbeefec\, 0\, 3\) \+ 2a0
>  00011b44 main     \(0\, 2d330\, 2cd08\, ffbeefec\, 0\, 0\) \+ 9c
>  00011a80 \_start   \(0\, 0\, 0\, 0\, 0\, 0\) \+ b8
> 

This has been fixed with change \#26043\.

> \./perl \-e 'sub bang \{ print "bang\!\\n"; \} undef %main&#8203;::; bang;'
bang\!

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From jeff@zeroclue.com


At least with perl 5\.8\.6 on Mac OS X Tiger\, this appers to still be a
problem&#8203;:

\[goo&#8203;:\~/Data Files\] jeff% perl \-e 'END\{ %&#8203;::=\(\) \}'
Bus error
\[goo&#8203;:\~/Data Files\] jeff% perl \-v

This is perl\, v5\.8\.6 built for darwin\-thread\-multi\-2level

**Comment by p5pRT** [comment]
### From jeff@zeroclue.com


At least with perl 5\.8\.6 on Mac OS X Tiger\, this appers to still be a
problem&#8203;:

\[goo&#8203;:\~/Data Files\] jeff% perl \-e 'END\{ %&#8203;::=\(\) \}'
Bus error
\[goo&#8203;:\~/Data Files\] jeff% perl \-v

This is perl\, v5\.8\.6 built for darwin\-thread\-multi\-2level
